src/ring_theory/ideal/cotangent.lean

```diff
@@ -0,0 +1,165 @@
+
+-/
+
+namespace ideal
+
+variables {R S : Type*} [comm_ring R] [comm_semiring S] [algebra S R] (I : ideal R)
+
+/-- `I ⧸ I ^ 2` as a quotient of `I`. -/
+@[derive [add_comm_group, module (R ⧸ I)]]
+def cotangent : Type* := I ⧸ (I • ⊤ : submodule R I)
+
+instance : inhabited I.cotangent := ⟨0⟩
+
+instance cotangent.module_of_tower : module S I.cotangent :=
+submodule.quotient.module' _
+
+instance : is_scalar_tower S R I.cotangent := by { delta cotangent, apply_instance }
+
+instance [is_noetherian R I] : is_noetherian R I.cotangent := by { delta cotangent, apply_instance }
+
+/-- The quotient map from `I` to `I ⧸ I ^ 2`. -/
+@[simps apply (lemmas_only)]
+def to_cotangent : I →ₗ[R] I.cotangent := submodule.mkq _
+
+lemma map_to_cotangent_ker : I.to_cotangent.ker.map I.subtype = I ^ 2 :=
+by simp [ideal.to_cotangent, submodule.map_smul'', pow_two]
+
+lemma mem_to_cotangent_ker {x : I} : x ∈ I.to_cotangent.ker ↔ (x : R) ∈ I ^ 2 :=
+begin
+  rw ← I.map_to_cotangent_ker,
+  simp,
+end
+
+lemma to_cotangent_eq {x y : I} : I.to_cotangent x = I.to_cotangent y ↔ (x - y : R) ∈ I ^ 2 :=
+begin
+  rw [← sub_eq_zero, ← map_sub],
+  exact I.mem_to_cotangent_ker
+end
+
+lemma to_cotangent_eq_zero (x : I) : I.to_cotangent x = 0 ↔ (x : R) ∈ I ^ 2 :=
+I.mem_to_cotangent_ker
+
+lemma to_cotangent_surjective : function.surjective I.to_cotangent :=
+submodule.mkq_surjective _
+
+lemma to_cotangent_range : I.to_cotangent.range = ⊤ :=
+submodule.range_mkq _
+
+lemma cotangent_subsingleton_iff :
+  subsingleton I.cotangent ↔ is_idempotent_elem I :=
+begin
+  split,
+  { introI H,
+    refine (pow_two I).symm.trans (le_antisymm (ideal.pow_le_self two_ne_zero) _),
+    exact λ x hx, (I.to_cotangent_eq_zero ⟨x, hx⟩).mp (subsingleton.elim _ _) },
+  { exact λ e, ⟨λ x y, quotient.induction_on₂' x y $ λ x y,
+      I.to_cotangent_eq.mpr $ ((pow_two I).trans e).symm ▸ I.sub_mem x.prop y.prop⟩ }
+end
+
+/-- The inclusion map `I ⧸ I ^ 2` to `R ⧸ I ^ 2`. -/
+def cotangent_to_quotient_square : I.cotangent →ₗ[R] R ⧸ I ^ 2 :=
+submodule.mapq (I • ⊤) (I ^ 2) I.subtype
+  (by { rw [← submodule.map_le_iff_le_comap, submodule.map_smul'', submodule.map_top,
+    submodule.range_subtype, smul_eq_mul, pow_two], exact rfl.le })
+
+lemma to_quotient_square_comp_to_cotangent : I.cotangent_to_quotient_square.comp I.to_cotangent =
+  (I ^ 2).mkq.comp (submodule.subtype I) :=
+linear_map.ext $ λ _, rfl
+
+@[simp]
+lemma to_cotangent_to_quotient_square (x : I) : I.cotangent_to_quotient_square (I.to_cotangent x) =
+  (I ^ 2).mkq x := rfl
+
+/-- `I ⧸ I ^ 2` as an ideal of `R ⧸ I ^ 2`. -/
+def cotangent_ideal : ideal (R ⧸ I ^ 2) :=
+{ carrier := (submodule.map (I ^ 2).mkq I).carrier,
+  add_mem' := λ _ _, (submodule.map (I  ^ 2).mkq I).add_mem,
+  zero_mem' := (submodule.map (I ^ 2).mkq I).zero_mem,
+  smul_mem' := begin
+    intros c x hx,
+    obtain ⟨c, rfl⟩ := ideal.quotient.mk_surjective c,
+    exact (submodule.map (I ^ 2).mkq I).smul_mem c hx,
+  end }
```

Comment:
This is definitionally nicer as
```suggestion
begin
  -- for some reason attempting to do this all at once makes the elaborator unhappy
  refine submodule.map _ I,
  rotate 2,
  have := @ring_hom.to_semilinear_map _ (R ⧸ I ^ 2) _ _,
  exact this (ideal.quotient.mk (I ^ 2)),
  refine ⟨quotient.surjective_quotient_mk'⟩,
end
```
which gives the term `submodule.map (quotient.mk (I ^ 2)).to_semilinear_map I`. I can't persuade the elaborator to accept this by itself though, hence the tactic mode.

Reply:
bors r-

bors d+